File: benchmark/synduce/combine/mss_with_sum.f
Inductive List = elt Int | cons {Int, List};
Inductive IdxList = ielt Int | icons {Int, Int, IdxList};

max = \a: Int. \b: Int. if < a b then b else a;

sum = fix (
  \f: IdxList -> Int. \xs: IdxList.
  match xs with
    ielt t -> t
  | icons {hd, idx, tl} -> + hd (f tl)
  end
);

hsum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt t -> t
  | cons {hd, tl} -> + hd (f tl)
  end
);

drop_sum_list = fix (
  \f: IdxList -> List. \xs: IdxList.
  match xs with
    ielt t -> elt t
  | icons {hd, idx, tl} -> cons {hd, f tl}
  end
);

mss = \xs: List. (fix (
  \f: List -> {Int, Int, Int}. \xs: List.
  match xs with
    elt t -> if (> t 0) then {t,t,t} else {0,0,0}
  | cons {hd, tl} ->
    let result = f tl in
    let mts_tl = result.1 in
    let mps_tl = result.2 in
    let mss_tl = result.3 in
    let sum_tl = hsum tl in
    let new_mps = max (+ hd mps_tl) 0 in
    let new_mts = max (+ hd sum_tl) mts_tl in
    {new_mts, new_mps, max new_mps mss_tl}
  end
) xs).3;

spec = \xs: List. mss xs;

repr = fix (
  \f: IdxList -> Compress List. \xs: IdxList.
  match xs with
    ielt t -> elt t
  | icons {hd, idx, tl} -> cons {hd, f tl}
  end
);

main = \xs: IdxList. spec (repr xs);

File: ml-benchmark/synduce/combine/mss_with_sum.ml
type list =
  | Elt of int
  | Cons of int * list

type idx_list =
  | IElt of int
  | ICons of int * int * idx_list

let max a b = if a > b then a else b

let rec sum = function
  | IElt x -> x
  | ICons (hd, idx, tl) -> hd + sum tl

let rec is_memo_sum = function
  | IElt x -> true
  | ICons (hd, idx, tl) -> idx == sum tl && is_memo_sum tl

let rec hsum = function
  | Elt x -> x
  | Cons (hd, tl) -> hd + hsum tl

let rec mss = function
  | Elt x -> if x > 0 then (x, x, x) else (0, 0, 0)
  | Cons (hd, tl) ->
    match mss tl with
    | (mts_tl, mps_tl, mss_tl) ->
      let sum_tl = hsum tl in
      let new_mps = max (mps_tl + hd) 0 in
      (max mts_tl sum_tl, new_mps, max new_mps mss_tl)

val drop_sum_list: idx_list -> list compress
let rec drop_sum_list = function
  | IElt x -> Elt x
  | ICons (hd, ids, tl) -> Cons (hd, drop_sum_list tl)

let prog xs = mss (drop_sum_list xs)
